### PR TITLE
Improved teamlist sorting

### PR DESCRIPTION
**Pull request by hettoo**
Using quicksort to sort the teamlists instead of the old naive approach allows removing the auxiliary lists, should be more efficient and saves some lines too.

In some gametypes less score might be better (like race, but not all race gametypes), so I added a gametype setting which indicates that the score ordering of those teamlists should be inverted.

To make the scoreboard look nicer I've made it order by name if scores are equal.
I just realized that this may not be desired, because it unnecessarily diverts the chase order from the scoreboard order even more, so you may want to remove line 76 and 77 from g_gameteams.c.

As an alternative I made this version which uses a list of all non-spectators, sorted the same way as the teamlists, to determine the order of chasing people: https://github.com/hettoo/qfusion/tree/playersorting2.
This could be nice, because you can estimate how many times you will have to click to reach a certain player.
Edit: changed from sorting a global playerlist to concatenating all sorted player teamlists.

Although this results in the order being changed when score differences change, I actually find it more convenient than 'randomly' wandering through the scoreboard.

### CODE REVIEW

**Comment by viciious** [comment]
What gametypes/modes should have the inverseScore property set to true?

2013/10/28 Gerco van Heerdt notifications@github.com

> Using quicksort to sort the teamlists instead of the old naive approach
> allows removing the auxiliary lists, should be more efficient and saves
> some lines too.
> 
> In some gametypes less score might be better (like race, but not all race
> gametypes), so I added a gametype setting which indicates that the score
> ordering of those teamlists should be inverted.
> 
> To make the scoreboard look nicer I've made it order by name if scores are
> equal.
> I just realized that this may not be desired, because it unnecessarily
> diverts the chase order from the scoreboard order even more, so you may
> want to remove line 76 and 77 from g_gameteams.c.
> 
> As an alternative I made this version which uses a list of all
> non-spectators, sorted the same way as the teamlists, to determine the
> order of chasing people:
> https://github.com/hettoo/qfusion/tree/playersorting2.
> This could be nice, because you can estimate how many times you will have
> to click to reach a certain player. Maybe instead of sorting all players
> together it should be a concatenation of the teamlists to make it even more
> predictable.
> 
> However, the problem with that alternative is that the order is not
> static. You could miss a person because his score changed some small time
> before you clicked to chase him. This problem is increased by the fact that
> 
> ## it takes some time before the scoreboard is updated at the client.
> 
> You can merge this Pull Request by running
> 
>   git pull https://github.com/hettoo/qfusion playersorting
> 
> Or view, comment on, or merge it at:
> 
>   https://github.com/viciious/qfusion/pull/34
> Commit Summary
> - More efficient teamlist sorting
> - Teamlist sort order: score, name, entity number
> - Add an inverse score gametype setting
> 
> File Changes
> - _M_ source/game/g_ascript.cpphttps://github.com/viciious/qfusion/pull/34/files#diff-0(1)
> - _M_ source/game/g_gameteams.cpphttps://github.com/viciious/qfusion/pull/34/files#diff-1(63)
> - _M_ source/game/g_gametypes.cpphttps://github.com/viciious/qfusion/pull/34/files#diff-2(2)
> - _M_ source/game/g_gametypes.hhttps://github.com/viciious/qfusion/pull/34/files#diff-3(1)
> 
> Patch Links:
> - https://github.com/viciious/qfusion/pull/34.patch
> - https://github.com/viciious/qfusion/pull/34.diff

## 

Best regards,
 Victor Luchitz

**Comment by hettoo** [comment]
For stock gametypes I thought race only, if only it would mirror the time of the player in his score like racesow, but it looks like it does not do that. :p
Could easily be added with something like a client.stats.setScore(this.finishTime) somewhere before line 190 in race.as, though.
Other uses would be gametypes where you start with a certain score and lose it on certain events or something like that, but afaik these do not exist at the moment.

Forgot to make it affect the score chasecam mode, added that now.

**Comment by hettoo** [comment]
Hm, actually the specific case of race is a little more complicated: a time of 0 is not the best possible time, but represents no time set yet. So instead of at the top, these times should end at the bottom.
I can either make this exception for gametypes with both the isRace and inverseScore flags set, or maybe  just drop the feature. What do you think?

**Comment by viciious** [comment]
It really depends on what player's score in race represents now. Is it the
best race time in milliseconds? Maybe it should just be changed to
something else, some number that would gracefully handle the unset times
too.

2013/10/30 Gerco van Heerdt notifications@github.com

> Hm, actually the specific case of race is a little more complicated: a
> time of 0 is not the best possible time, but represents no time set yet. So
> instead of at the top, these times should end at the bottom.
> I can either make this exception for gametypes with both the isRace and
> inverseScore flags set, or maybe just drop the feature. What do you think?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/viciious/qfusion/pull/34#issuecomment-27381309
> .

## 

Best regards,
 Victor Luchitz

**Comment by hettoo** [comment]
Ah right, didn't think of it that way. Projecting absent times as a score of MAX_INT should work.

Currently basewsw race does not set scores to anything as far as I can see. It does keep an array with data regarding checkpoints and finish times for all players.
The score could easily be kept mirrored to the time (in milliseconds indeed), if it would be updated when a player makes a new record, connects or changes team. No rush with that though, since nobody actually plays stock race and this feature won't change any default behavior anyway.

Racesow race will be using it next version. Although that is a mod which could add it itself, I guess this is generic enough to be included here.

File: source/game/g_gameteams.cpp [hunk 1]
```diff
@@ -66,6 +66,20 @@ void G_Teams_Init( void )
 
 }
 
+static int G_Teams_CompareMembers( const void *a, const void *b )
+{
+	edict_t *edict_a = game.edicts + *(int *)a;
+	edict_t *edict_b = game.edicts + *(int *)b;
+	int score_a = edict_a->r.client->level.stats.score;
+	int score_b = edict_b->r.client->level.stats.score;
+	int result = ( level.gametype.inverseScore ? -1 : 1 ) * ( ( score_b > score_a ) - ( score_a > score_b ) );
```
Comment (viciious) [inline source/game/g_gameteams.cpp]: Why not:
int result = ( level.gametype.inverseScore ? -1 : 1 ) \* ( score_b - score_a );

File: source/game/g_gameteams.cpp [hunk 2]
```diff
@@ -66,6 +66,20 @@ void G_Teams_Init( void )
...
+	if (!result)
+		result = Q_stricmp( edict_a->r.client->netname, edict_b->r.client->netname );
+	if (!result)
+		result = ( ENTNUM( edict_b ) < ENTNUM( edict_a ) ) - ( ENTNUM( edict_a ) < ENTNUM( edict_b ) );
```
Comment (viciious) [inline source/game/g_gameteams.cpp]: Same as with line #75

**Comment by hettoo** [comment]
I must have seen some example doing it that way and assumed the output is supposed to be either -1, 0 or 1. The documentation does not say so, so I'll change it.

**Comment by viciious** [comment]
Quoting
http://www.gnu.org/software/libc/manual/html_node/Array-Sort-Function.html :

The compare function is used to perform the comparison on the array
elements. This function is called with two pointer arguments and should
return an integer less than, equal to, or greater than zero corresponding
to whether its first argument is considered less than, equal to, or greater
than its second argument.

It never says it should return -1, 0 or 1.

2013/10/30 Gerco van Heerdt notifications@github.com

> I must have seen some example doing it that way and assumed the output is
> supposed to be either -1, 0 or 1. The documentation does not say so, so
> I'll change it.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/viciious/qfusion/pull/34#issuecomment-27419271
> .

## 

Best regards,
 Victor Luchitz

**Comment by viciious** [comment]
Is there a way to make the chase order match the new sorting algorithm?

**Comment by hettoo** [comment]
Yes, with these additional changes: https://github.com/hettoo/qfusion/compare/playersorting...playersorting2
It keeps a global playerlist with all sorted non-spectator teams next to each other and uses that for chasing. So as a spectator in a team-based game you first traverse alpha in order, then beta.

There is a gitignore commit above because I merged with my master of which I forgot I added that after the new angelscript directory. ^^

**Comment by viciious** [comment]
I am not sure which pull request I am supposed to merge now :(

2013/11/6 Gerco van Heerdt notifications@github.com

> Yes, with these additional changes: hettoo@playersorting...playersorting2https://github.com/hettoo/qfusion/compare/playersorting...playersorting2
> It keeps a global playerlist with all sorted non-spectator teams next to
> each other and uses that for chasing. So as a spectator in a team-based
> game you first traverse alpha in order, then beta.
> 
> There is a gitignore commit above because I merged with my master of which
> I forgot I added that after the new angelscript directory. ^^
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/viciious/qfusion/pull/34#issuecomment-27888275
> .

## 

Best regards,
 Victor Luchitz

**Comment by hettoo** [comment]
Hehe. I prefer having that change as well, so I've merged that to this request.
Merge this if you agree with everything. :)

**Comment by viciious** [comment]
A few minor things:

because of this:

playerlist[playerlist_numplayers] = -1;

the playerlist variable needs to be declared as:

int playerlist[MAX_CLIENTS+1];

not as:

int playerlist[MAX_CLIENTS];

Also, I don't like the naming of the variable, it's too vague.

2013/11/6 Gerco van Heerdt notifications@github.com

> Hehe. I prefer having that change as well, so I've merged that to this
> request.
> Merge this if you agree with everything. :)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/viciious/qfusion/pull/34#issuecomment-27896649
> .

## 

Best regards,
 Victor Luchitz

File: source/game/g_gametypes.cpp [hunk 3]
```diff
@@ -21,6 +21,8 @@
 #include "g_local.h"
 
 g_teamlist_t teamlist[GS_MAX_TEAMS];
+int playerlist[MAX_CLIENTS];
```
Comment (viciious) [inline source/game/g_gametypes.cpp]: Has to be MAX_CLIENTS+1 because of:
playerlist[playerlist_numplayers] = -1;

**Comment by hettoo** [comment]
True, and the same goes for the playerIndices arrays in the teamlists.
Actually this -1 ending should not be needed, since the sizes of the arrays are tracked as well.

**Comment by viciious** [comment]
Now if only you renamed playerlist to something more descriptive..

2013/11/7 Gerco van Heerdt notifications@github.com

> True, and the same goes for the playerIndices arrays in the teamlists.
> Actually this -1 ending should not be needed, since the sizes of the
> arrays are tracked as well.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/viciious/qfusion/pull/34#issuecomment-27963113
> .

## 

Best regards,
 Victor Luchitz

**Comment by hettoo** [comment]
Not sure about the name: something like sorted_players may imply a globally sorted list, while it is only a concatenation of sorted teamlists.
Perhaps the full list should not be stored at all, considering it will probably only be used for this, but the chase step code should loop over all individual teamlists in turn instead.

**Comment by viciious** [comment]
So is this pull request ok to be merged now?

2013/11/7 Gerco van Heerdt notifications@github.com

> Not sure about the name: something like sorted_players may imply a
> globally sorted list, while it is only a concatenation of sorted teamlists.
> Perhaps the full list should not be stored at all, considering it will
> probably only be used for this, but the chase step code should loop over
> all individual teamlists in turn instead.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/viciious/qfusion/pull/34#issuecomment-27966816
> .

## 

Best regards,
 Victor Luchitz

**Comment by hettoo** [comment]
Yeah, go ahead. :)

**Comment by hettoo** [comment]
Found a little missing statement in that last change: if the player you chase disappeared the team loop variable isn't reset.
This causes no error though, presumably some other function picking another player when that happens, but it would be nice if you could add an "if( team == GS_MAX_TEAMS ) team = TEAM_PLAYERS;" before line 507 in g_chase.c, just in case.
Thanks and sorry for this.

**Comment by viciious** [comment]
Please make another pull request, I'm afraid I can break something there
since I don't quite understand the code. Also, some code comments would be
nice too :)

2013/11/7 Gerco van Heerdt notifications@github.com

> Found a little missing statement in that last change: if the player you
> chase disappeared the team loop variable isn't reset.
> This causes no error though, presumably some other function picking
> another player when that happens, but it would be nice if you could add an
> "if( team == GS_MAX_TEAMS ) team = TEAM_PLAYERS;" before line 507 in
> g_chase.c, just in case.
> Thanks and sorry for this.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/viciious/qfusion/pull/34#issuecomment-27982724
> .

## 

Best regards,
 Victor Luchitz

**Pull request merged**